### PR TITLE
Enable universal (mono + MS.Net) support using the same asembly files.

### PR DESCRIPTION
**Pull request by pruiz**
Hi,

Actual mono support requires manually re-building SE.Redis using mono's mcs compiler. Something a little bit painfull when you are depending on automated buils using NuGet dependencies, or trying to achieve universal deployments of the same binaries across platforms.

In order to avoid having to maintain an specific build projects for mono, I've removed conditional compilation based on "_MonoCS_" directive, and swapped it with a simple MonoHelper providing a couple of dynamic wrappers in order to allow the same assembly file to be used on both Win & Mono targets.

This should allow maintainers of software (or NuGets) depending on StackExchange to avoid having to release an specific NuGet package using ad-hoc packaged StackExchange.Redis.mono assemblies.

NOTE: I've tried to maintain coding style, etc. but just let me know anything which may be missing.. ;)

### CODE REVIEW

**Comment by mgravell** [comment]
I haven't reviewed yet, but most (if not all) of those conditionals are
there to work around the lack of **highly desirable** tools available on
some platforms. I am rather sceptical of any attempt to remove them, as
that suggests (and again, I haven't reviewed yet) some very significant and
undesirable compromises.

I would much rather support a new mono build in the nuget... which should
be pretty easy when we've merged the Core-CLR work.

**Comment by pruiz** [comment]
Well, I know (read: I've suffered) the mono limitations/issues when it comes to network sockets. However I've not tried to modify the current SR.Redis behavior. I've simply tried to make a single assembly which dynamically uses one implementation (poll vs async) depending on the platform it is are actually running on.

That said, your proposal of having a mono-specific build, while probably usefull for others, makes it really hard to provide universal builds (ie. same binaries deployables to both ms.net and mono), as the SR.Redis dll which is to be packaged as deployment unit would be chosen at build time.. :(

**Comment by mgravell** [comment]
Agreed, that is a vexing limitation. Will review with that in mind. Note
that this is likely to have merge conflicts with the pending core-clr work
:(

**Comment by pruiz** [comment]
Don't worry, I can rebase this against core-clr changes if that makes it easier.. And I was expecting comments and/or corrections arising from code review.. (In fact, teviewing the pull-req I've already spotted a couple of leftovers I will try to address tonite. ;))

**Comment by pruiz** [comment]
@mgravell  Did you had any chance of reviewing this? (Or maybe you are waiting until core-clr changes are merged into master?)

**Comment by mgravell** [comment]
I've been merging the other PR into the CoreCLR branch while it stabilises.
Did a deploy of that yesterday. Haven't even read this yet, due to time
limits. Very likely that it will be awkward to merge the two, but: we'll
see!

**Comment by pruiz** [comment]
@mgravell Do you have any updates on this? ;)

I understand you are right now working on CoreCLR migration, but please, let me know if I can help by trying to provide a version of this patch against CoreCLR branch or something.

**Comment by mgravell** [comment]
I've merged the CoreCLR stuff - which removes the barriers to merging this; you indicated:

> let me know if I can help by trying to provide a version of this patch against CoreCLR branch or something.

If you are willing to update to the current master, I'd love to look at it. Otherwise, I can try to merge it, but you're much more familiar with the changes you've made.

**Comment by pruiz** [comment]
I'll try to rebase this patch same week, I'll let you know then. ;)

**Comment by NickCraver** [comment]
@pruiz any chance you'll have time for this? You're in a much better position to test such changes against Mono, so we'd still very much appreciate a PR update here.

**Comment by pruiz** [comment]
I am a little bit busy right now, but I'll try to take a look at this
during this week. ;)

**Comment by pruiz** [comment]
Hi @NickCraver,

I've just rebased & updated this changes, and tested against mono. However I am still trying to get howto build the core-clr part.. ;)

**Comment by pruiz** [comment]
Hi @NickCraver / @mgravell,

Did you had a chance on looking at this?

**Comment by mgravell** [comment]
Oh, awesome; I didn't see the rebase arrive. Thanks, will review.

**Comment by kotharin** [comment]
Any update on when this will be merged? released?

**Comment by mgravell** [comment]
When I've had chance to go through it; sorry, but I've got a lot of plates
spinning at the moment. I'm not intentionally delaying, but I also don't
want to over-promise.

On 26 February 2016 at 16:06, kotharin <notifications@github.com> wrote:

> Any update on when this will be merged? released?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/pull/314#issuecomment-189343707>
> .
>



-- 
Regards,

Marc

**Comment by kotharin** [comment]
i forked the branch that pruiz created. I was able to connect to redis on linux using mono, when I compiled the dll using mono. If I compiled the dll in windows and tried to use it on linux with mono, it didn't work. Not sure if this is the expected behavior.

**Comment by mgravell** [comment]
Can you be more specific about "didn't work"? I doubt that is the intended
behaviour.

**Comment by pruiz** [comment]
How did hoy exactly build the assembly on each case?

Take into account that the whole purpose of this patch is to avoid having different builds. So there should be no difference (from a functional/behavior  point of view) from building on Linux (using xbuild+mono) or on Windows (msbuild+dotnet).

 In both cases any of the generated assemblies should work on both platforms (albeit changing some of its internal runtime behavior depending on where the code is actually being executed).

Also I specifically removed conditional compilation defines regarding mono/*ix vs dotnet/mswin, as to have the same building logic on both environments.

Regarding the case where it didn't work:

 - do you have any kind of error/exception or something? 

 - can you provide some details regarding the building and runtime environment of your tests?

 - did you try to execute the tests already available at stackexchange.redis solution? Or were you running some sort of project/application of your own?

**Comment by kotharin** [comment]
I use the netbuild.cmd and the monobuild.bash in the appropriate environments to build the library. When I tested with the windows build, on linux using mono and a custom app. I get the following error:
Unhandled Exception:
StackExchange.Redis.RedisConnectionException: It was not possible to connect to                                                            the redis server(s); to create a disconnected multiplexer, disable AbortOnConnec                                                           tFail. UnableToResolvePhysicalConnection on PING
  at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl (System.Func`1 multip                                                           lexerFactory, System.IO.TextWriter log) <0x41e85d90 + 0x0011f> in <filename unkn                                                           own>:0
  at StackExchange.Redis.ConnectionMultiplexer.Connect (System.String configurat                                                           ion, System.IO.TextWriter log) <0x41e854e0 + 0x00117> in <filename unknown>:0
  at Marvel.Redis.RedisModule.connDb (System.String host) <0x41e84780 + 0x0000f> 

The app works fine when using the library from the mono build. 

The weird thing is that when I run the BasicTest.exe using mono with either library, I get the following error:

Unhandled Exception:
StackExchange.Redis.RedisConnectionException: It was not possible to connect to the redis server(s); to create a disconnected multiplexer, disable AbortOnConnectFail. UnableToResolvePhysicalConnection on PING
  at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl (System.Func`1 multiplexerFactory, System.IO.TextWriter log) <0x4165daf0 + 0x0011f> in <filename unknown>:0
  at StackExchange.Redis.ConnectionMultiplexer.Connect (System.String configuration, System.IO.TextWriter log) <0x4165d590 + 0x00103> in <filename unknown>:0
  at BasicTest.Program.MassiveBulkOpsAsync (Int32 AsyncOpsQty, Boolean preserveOrder, Boolean withContinuation) <0x4165cf30 + 0x00073> in <filename unknown>:0
  at BasicTest.Program.Main (System.String[] args) <0x4165cd50 + 0x00063> in <filename unknown>:0
[ERROR] FATAL UNHANDLED EXCEPTION: StackExchange.Redis.RedisConnectionException: It was not possible to connect to the redis server(s); to create a disconnected multiplexer, disable AbortOnConnectFail. UnableToResolvePhysicalConnection on PING

**Comment by mgravell** [comment]
My stretch goal for this week is to get this reviewed; sorry, lots of changes to work through, and I've been *insanely* busy on 20 things

**Comment by pruiz** [comment]
Hi Marc,

I've been a bit busy too lately, but just let me know if you have any
issues.

Regards

**Comment by pruiz** [comment]
@mgravell Hi, did you had a chance to review this?

**Comment by dobryanskyy** [comment]
Hi @pruiz , @mgravell ,

We're also waiting for this, cause we have nuget restoring our packages both on mono(linux) and ms.net(windows). 
So keeping fingers crossed, good luck in merging. :+1: -)

**Comment by pruiz** [comment]
@mgravell  ping? ;)

**Comment by pruiz** [comment]
@mgravell Hi, I've just upgraded this pull-requests, and tested it (using BasicTest, etc.) against both linux (using mono 4.2.1 & 4.6.1), windows 10 (using .net 4.6). Using always the same binaries (originally built on windows, using VS 2k15).

I hope you can find some time to have a look at this pull-request, as this has been laying around nearly for a year now.. ;)

**Comment by pruiz** [comment]
@NickCraver, I am pinging you, as you were interested in this pull-req too. ;)

**Comment by voidmonk** [comment]
Hey guys, any luck with reviewing and pulling this in. It's one of those deal-breaker issues when trying to get things running cross-platform in a friction-less automated way. A single assembly (for Windows/.NET and Linux/Mono) is an ideal solution.

**Comment by NickCraver** [comment]
I'm looking at everything in queue here - and for Mono and more specifically `netstandard` (and VS 2017), this one's something we need to get in before a move. I am a little concerned about the helper behavior in 2 specific ways:

1. It's a try/catch each time, rather than remembering `X` doesn't work on the current platform.
2. Are there no platform bits available in the netstandard space to us so we can determine what's there or not at runtime, rather than a try/fail/try again pattern?

Minor: in the SslStream bits, `result` will always be null.

**Comment by pruiz** [comment]
With regard to 1: I will rework it so the initial action get's cached, no prob :)
With regard to 2: I ma not understand your point here, but I am afraid there's no specific API available in order to ask the runtime wether this stuff/api/method is available or not, except asking wether we are running under mono. But even so, some of this stuff (like the ssl issues) maybe implemented/supported/fixed at any upcoming mono version, and my intention was to provide a way which would work transparently if/when running on such future-mono-version supporting it.

I'll try to take a look at this during the weekend and be back here..

File: StackExchange.Redis/StackExchange/Redis/MonoHelper.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		public static bool RunninOnUnix
+		{
+			get
+			{
+				int p = (int)Environment.OSVersion.Platform;
+				return ((p == 4) || (p == 6) || (p == 128));
+			}
+		}
+
+		public static bool RunninOnLinux
```
Comment (redoz) [inline StackExchange.Redis/StackExchange/Redis/MonoHelper.cs]: This also looks like a typo, should it be "RunningOnLinux"?

**Comment by grinsted** [comment]
@pruiz Could something like this be used:

```c#
		private static Nullable<bool> cachedSslStreamTakesEncryptionPolicy = null;

		private static bool SslStreamTakesEncryptionPolicy()
		{
			if (cachedSslStreamTakesEncryptionPolicy == null)
			{
				Type t = typeof(SslStream);
				foreach (ConstructorInfo ci in (t.GetConstructors()))
				{
					foreach (ParameterInfo pi in ci.GetParameters())
					{
						Console.WriteLine(pi.ParameterType.FullName);
						if (pi.ParameterType.FullName == "System.Net.Security.EncryptionPolicy")
						{
							cachedSslStreamTakesEncryptionPolicy = true;
							return true;
						}
					}
				}
				cachedSslStreamTakesEncryptionPolicy = false;
			}
			return (bool)cachedSslStreamTakesEncryptionPolicy;
		}
```

**Comment by pruiz** [comment]
Hi Aslak,

Probably yes, I've been just to busy lately to work on this.. I hope I can
catchup soon, thought.

**Comment by NickCraver** [comment]
The project structure has changed a ton for `netstandard`, and I'm honestly just not that familiar with `netstandard` and how well Mono supports it. So excuse the possibly very naive question, but: is this still needed, or do the current builds work properly?

**Comment by pruiz** [comment]
My guess is probably yes..  But I'll try it and report back here.

**Comment by pruiz** [comment]
Hi @NickCraver,

I am afraid this is still an issue with latest master, as simply trying BasicTest under mono/osx fails miserably with:  It was not possible to connect to the redis server(s); to create a disconnected multiplexer, disable AbortOnConnectFail.

:(

**Comment by emitelma** [comment]
Why can't you guys fix this? This is ridiculous.

**Comment by Salgat** [comment]
Any news on when this will get merged or at least reviewed again?

**Comment by ondravondra** [comment]
Will this ever be merged?

**Comment by 564064202** [comment]
how to resolve this ?

**Comment by shravan2x** [comment]
@mgravell Any ETA?

**Comment by mgravell** [comment]
"whenever I can get the time to do it". Right now all my time is taken
preparing for redisconf. Hopefully I should get some after that...

**Comment by NickCraver** [comment]
An update for those here: this PR is *drastically* out of date - we don't know what the current issues with Mono are and don't use it daily (so have to do a lot of setup here). The PR itself targets the old .csproj system not used in forever, and takes away polling approaches used by *most* people right now for the sake of compatibility. This in itself is very problematic and likely results in a broken library (especially under load) for most.

If anyone wants to take a crack at a fresh PR we could test on Mono that'd be awesome. It's not our area of expertise. I wanted to give an update here though, since it's obviously a high area of community interest, unfortunately it's not also an area of community investment...so we have to find the (great deal of time) to really analyze this from the ground up.

It'd really help knowing what *breaks* on Mono as a starting point. Unfortunately my MacBook's out of commission now, so recommendations on the best way to test Mono would also be appreciated. At the moment, I'd love to know what the gap in `netstandard2.0` build that we have today and Mono is...that's possibly a much smaller set of things to address than this PR aimed at.

**Comment by shravan2x** [comment]
@NickCraver I don't have the stack trace exactly, but there's some issue with the connection manager which prevents the client from even connecting to the server.

**Comment by emitelma** [comment]
Is redis driver ever going to work with Mono? This is silly...

**Comment by Salgat** [comment]
@NickCraver Is it safe to say that if someone puts in the effort again, that this time around it will be merged in?

**Comment by NickCraver** [comment]
@Salgat if it doesn't have the huge downsides mentioned in my comment (breaking the other platforms), yes. It's a lot easier to take a look at the problem now with the new project system, but my MacBook has also died. Will need to be able to test somewhere (we can't on our current build server). I'll try to get setup for Mono testing on a VM here, but I still don't know all of the problems with the current lib on Mono. Connections are step 1, but what else is broken I'm unsure.

The kicker to this is with the Sockets/Pipelines combo that @mgravell and I have been waiting on (I don't think this is happening in the BCL any time soon - so we'll have to write it) may resolve the Mono issues entirely without any special code, due to the drastic sockets improvements in .NET Core 2.1 in general. With these in play, we can drastically simplify the connection code and (hopefully) resolve the Mono issues at the same time. Though this would likely be a `netstandard2.0` (Mono 5.4) minimum bar for this. Given all the scenarios here, I think that minimum bar is most likely.

I'm curious what those here think of that version minimum, though.

**Comment by shravan2x** [comment]
@NickCraver Given how much larger the API surface in `netstandard2.0` is than previous versions, I think that's very reasonable. Most people are likely using it anyway.

**Comment by NickCraver** [comment]
Hmmm, I need some additional info needed here. I've just installed Mono 5.12 on Windows 10 and unfortunately I'm not hitting any fundamental issues. Running our test suite:
```
  Finished:    StackExchange.Redis.Tests
=== TEST EXECUTION SUMMARY ===
   StackExchange.Redis.Tests  Total: 825, Errors: 0, Failed: 8, Skipped: 32, Time: 204.682s
```

The few failures are known jitter in the suite (same as the build) with races we still need to iron out. For running the full suite this is nor abnormal at the moment. I was hoping to get complete failure on all connections given the posts in this issue.

So - what are people using? Which OS and which versions of Mono? Is it Linux, macOS? (I hope not...I have no working Mac) This info would help on where to prod next.

**Comment by pruiz** [comment]
@NickCraver The main issue was the sockets/connecting poll vs no-poll one. Everything else was minor and depended on the minimun mono version used:

- There is the "CompressionLevel.Optimal" issue, which at least at the time, was not implemented on mono.
- Then there was the Certificate Validation callbacks support, which may have been fixed in recent mono versions. But take into account that every major mono version since 4.8 has updated/reworked the inner SSL/TLS support in some manner. So "there be dragons" may apply here.. ;)

And that was it all back at the time.

**Comment by shravan2x** [comment]
@NickCraver I had connection issues on Linux. I think the Mono version was 5.4, but I don't remember.

**Comment by NickCraver** [comment]
I've taken a stab at this, but it's not that simple to support, and I wanted to give an update here after spending all afternoon on it.

First, I've enabled the test suite to run on Mono. It runs fine on Windows (passing as it should) and runs but fails most of the test on *nix (expected). Luckily WSL makes this a lot easier to build and test quickly. The commit fixing tests so they can run is here: b987ab3ff3a66ae9eca48fc5a33f4dfdfd0e107c

If you want to run it on WSL, after you're setup with Mono, here's my example command
```
/mnt/c/git/StackExchange/StackExchange.Redis/StackExchange.Redis.Tests/bin/Debug/net462$
mono /mnt/c/Users/myUser/.nuget/packages/xunit.runner.console/2.4.0-beta.2.build3981/tools/net452/xunit.console.exe StackExchange.Redis.Tests.dll
```

Now, when looking through the changes here, the platform helpers, and what's available, I've come to a decision there's no quick way around: we have to basically choose between `netstandard1.x` (which lacks the platform identifiers to identify Mono correctly), or Mono. There are several splits around what's supported on the socket side and the platform side (e.g. `Environment.OSVersion.Platform` is not available). Or, we have to add a dependency on the platform packages, just for Mono. None of these are good options.

Honestly, after going through the socket splits and the changes to networking we want to make, my recommendation is to drop `netstandard1.5` support, which would make things easier, but I'm not sure what impacts the most users as we have no visibility into those numbers.

So some questions:
- Who here is still using Mono?
- What's the motivation for Mono over .NET Core at this point?
- Do you see yourself moving?

I'll be blunt: it'd be nice to support everyone, but if the cost and complexity is non-trivial, we have to choose based on how many people it'll help vs. the cost. I'm very unsure about that tradeoff in this case. Answers to the above and counts would help us a bit here.

**Comment by pruiz** [comment]
One here using mono, with projects with dependencies not supporting netcore/netstandard. So I’ll vote for supporting mono instead of netstandard 1.5.. ;)

**Comment by Salgat** [comment]
I'm curious what difference it makes whether people answer those questions. We know users are stuck with .NET Framework for many reasons (porting over projects, especially without compatible dependencies is a massive undertaking) and we know they'd like to use this library with Mono considering the large interest in this PR and related issues. It goes without saying.

**Comment by emitelma** [comment]
Yes. It simple does not work on Mono and we're stuck with v1.0 build.
That's very unfortunate.

**Comment by NickCraver** [comment]
@Salgat There’s some interest, yes, but in terms on the user base for this library, it’s not large. It’s about a dozen people over 2.5 years (some of those may have already moved on, hence the questions) and a handful more in a few issues. These numbers obviously are just representative and not absolute by any stretch of the imagination, but given the Mono solution has been “build it yourself” since the beginning, I’m not sure how many users this group represents.

I am trying to see if we can reasonably do this (see #842 and #843), but can’t make any promises here. I don’t even have it working yet. If not, I would like to restore the build option at a bare minimum, if we can.

**Comment by voidmonk** [comment]
I waited a long time to be able to use this library for our startup product, hoping to host with Mono on Linux in production, but moved-on last year to continue to use .NET Framework on Windows for now, and planning to eventually migrate to .NET Core. Also had doubts about Mono's performance and future road-map.

**Comment by shravan2x** [comment]
@NickCraver I think Mono support is more important since `netstandard1.x` users can upgrade to `netstandard2.0` anyway.

**Comment by NickCraver** [comment]
I've ported the changes into the current state of the world (I think), but still can't get this working on Mono: If anyone's able to give it a shot or help debug, #845 has the code up. I've been unable to get it working thus far. It *runs*, but won't successfully connect.

**Comment by pruiz** [comment]
I've tried building on vs4mac (mono 5.10), but it fails to build due to nuget restoring issues.. I'll give it a try again later today and post here.

**Comment by pruiz** [comment]
@NickCraver finally got to the point of running the test suite using xunit, etc. but I have the same failures as you had. Weird, cause 

```
BayMax:StackExchange.Redis pruiz$ mono ./xunit.runner.console.2.4.0-beta.2.build3981/tools/net452/xunit.console.exe ./StackExchange.Redis.Tests/bin/Debug/net462/StackExchange.Redis.Tests.dll
xUnit.net Console Runner v2.4.0-beta.2.build3981 (64-bit Desktop .NET 4.0.30319.42000)
  Discovering: StackExchange.Redis.Tests
  Discovered:  StackExchange.Redis.Tests
  Starting:    StackExchange.Redis.Tests
    Bits.BasicOps [FAIL]
      System.TimeoutException : Connect timeout
      Stack Trace:
          at StackExchange.Redis.Tests.TestBase.Create (System.String clientName, System.Nullable`1[T] syncTimeout, System.Nullable`1[T] allowAdmin, System.Nullable`1[T] keepAlive, System.Nullable`1[T] connectTimeout, System.String password, System.String tieBreaker, System.IO.TextWriter log, System.Boolean fail, System.String[] disabledCommands, System.String[] enabledCommands, System.Boolean checkConnect, System.Boolean pause, System.String failMessage, System.String channelPrefix, System.Boolean useSharedSocketManager, System.Nullable`1[T] proxy) [0x001fa] in <4db2f3a2b144415589a228a60dca2e55>:0
          at StackExchange.Redis.Tests.Bits.BasicOps () [0x00001] in <4db2f3a2b144415589a228a60dca2e55>:0
          at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
          at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in <bb7b695b8c6246b3ac1646577aea7650>:0
      Output:
        Compiled under .NET 4.6.2
        Failed to get OS version
        10:25:34.123: 127.0.0.1:6379,127.0.0.1:6380,password=*****
        10:25:34.131:
        10:25:34.131: Connecting 127.0.0.1:6379/Interactive...
        10:25:34.131: BeginConnect: 127.0.0.1:6379
        10:25:34.131: Connecting 127.0.0.1:6380/Interactive...
        10:25:34.131: BeginConnect: 127.0.0.1:6380
        10:25:34.131: EndConnect: 127.0.0.1:6379
        10:25:34.132: Connected Interactive/127.0.0.1:6379
        10:25:34.132: Server handshake
        10:25:34.132: Setting client name: BayMax
        10:25:34.132: Auto-configure...
        10:25:34.132: Sending critical tracer: Interactive/127.0.0.1:6379
        10:25:34.132: Writing to Interactive/127.0.0.1:6379: ECHO
        10:25:34.132: Flushing outbound buffer
        10:25:34.132: Starting poll
        10:25:34.132: 2 unique nodes specified
        10:25:34.132: Requesting tie-break from 127.0.0.1:6379 > __Booksleeve_TieBreak...
        10:25:34.132: EndConnect: 127.0.0.1:6380
        10:25:34.132: Connected Interactive/127.0.0.1:6380
        10:25:34.132: Server handshake
        10:25:34.132: Setting client name: BayMax
        10:25:34.132: Auto-configure...
        10:25:34.132: Sending critical tracer: Interactive/127.0.0.1:6380
        10:25:34.132: Writing to Interactive/127.0.0.1:6380: ECHO
        10:25:34.132: Flushing outbound buffer
        10:25:34.132: Connect complete: 127.0.0.1:6379
        10:25:34.132: Starting poll
        10:25:34.132: Connect complete: 127.0.0.1:6380
        10:25:34.134: Requesting tie-break from 127.0.0.1:6380 > __Booksleeve_TieBreak...
        10:25:34.135: Allowing endpoints 00:00:05 to respond...
        10:25:34.135: Awaiting task completion, IOCP: (Busy=0,Free=200,Min=4,Max=200), WORKER: (Busy=0,Free=400,Min=4,Max=400)
        10:25:39.136: Not all tasks completed cleanly, IOCP: (Busy=0,Free=200,Min=4,Max=200), WORKER: (Busy=4,Free=396,Min=4,Max=400)
        10:25:39.136: 127.0.0.1:6379 did not respond
        10:25:39.136: 127.0.0.1:6380 did not respond
        10:25:39.136: Awaiting task completion, IOCP: (Busy=0,Free=200,Min=4,Max=200), WORKER: (Busy=4,Free=396,Min=4,Max=400)
        10:25:39.186: Not all tasks completed cleanly, IOCP: (Busy=0,Free=200,Min=4,Max=200), WORKER: (Busy=1,Free=399,Min=4,Max=400)
        10:25:39.186: 127.0.0.1:6379 failed to nominate (WaitingForActivation)
        10:25:39.186: 127.0.0.1:6380 failed to nominate (WaitingForActivation)
        10:25:39.186: No masters detected
        10:25:39.188: 127.0.0.1:6379: Standalone v2.0.0, master; keep-alive: 00:01:00; int: ConnectedEstablishing; sub: ConnectedEstablishing; not in use: DidNotRespond
        10:25:39.190: 127.0.0.1:6379: int ops=7, qu=2, qs=7, qc=0, wr=0, socks=1; sub ops=3, qu=0, qs=3, qc=0, wr=0, socks=1
        10:25:39.195: Circular op-count snapshot; int: 0+7=7 (0.70 ops/s; spans 10s); sub: 0+3=3 (0.30 ops/s; spans 10s)
        10:25:39.195: 127.0.0.1:6380: Standalone v2.0.0, master; keep-alive: 00:01:00; int: ConnectedEstablishing; sub: ConnectedEstablishing; not in use: DidNotRespond
        10:25:39.195: 127.0.0.1:6380: int ops=7, qu=2, qs=7, qc=0, wr=0, socks=1; sub ops=3, qu=0, qs=3, qc=0, wr=0, socks=1
        10:25:39.195: Circular op-count snapshot; int: 0+7=7 (0.70 ops/s; spans 10s); sub: 0+3=3 (0.30 ops/s; spans 10s)
        10:25:39.196: Sync timeouts: 0; fire and forget: 0; last heartbeat: -1s ago
        10:25:39.196: resetting failing connections to retry...
        10:25:39.197: retrying; attempts left: 2...
        10:25:39.197: 2 unique nodes specified
        10:25:39.197: Requesting tie-break from 127.0.0.1:6379 > __Booksleeve_TieBreak...
        10:25:39.197: Requesting tie-break from 127.0.0.1:6380 > __Booksleeve_TieBreak...
        10:25:39.197: Allowing endpoints 00:00:05 to respond...
        10:25:39.197: Awaiting task completion, IOCP: (Busy=0,Free=200,Min=4,Max=200), WORKER: (Busy=4,Free=396,Min=4,Max=400)
```

**Comment by pruiz** [comment]
@NickCraver, here is an update on this: From what I can see, somehow PlatformHelper.DefaultSocketMode ends up with 'Poll' value (instead of Async, which is the mode supported on mono). Hardcoding it to Async allowed me to run the test suite (albeit with failures).

I'll need to dig a bit deeper later this week in order to check wether this problem comes from my working copy or environment.. :?

**Comment by pruiz** [comment]
This is the final results from my test suite run on OSX (w/ mono 5.10.1.47)

```
  Finished:    StackExchange.Redis.Tests
=== TEST EXECUTION SUMMARY ===
   StackExchange.Redis.Tests  Total: 824, Errors: 0, Failed: 13, Skipped: 32, Time: 266.418s
```

**Comment by pruiz** [comment]
@NickCraver I've uploaded my changes to https://github.com/pruiz/StackExchange.Redis/tree/pull-845, I'll be glad if you can give it a try or review the changes..

This code is able to run the testsuite on mono/osx, giving just 13 failures.. which I will need to review next.

Had no chance of testing a dll built with this changes on ms.net (I have no windows computer here now), but I will try it tomorrow.

**Comment by pruiz** [comment]
@NickCraver I am running the test suite with the following command, using mono 5.10.1.47 (this is the "stable" version used by vs4mac right now)..
```
mono ./xunit.runner.console.2.4.0-beta.2.build3981/tools/net452/xunit.console.exe ./StackExchange.Redis.Tests/bin/Debug/net462/StackExchange.Redis.Tests.dll
```

Maybe you can try downgrading to 5.10.1.47 using mono repository packages for ubuntu: https://www.mono-project.com/download/vs/#download-lin-ubuntu, and apt-get install mono=5.10.1.47-0xamarin4+ubuntu1804b1 (or equivalent version tag for the ubuntu version used by WSL)

**Comment by NickCraver** [comment]
@pruiz Couldn't downgrade, had to pave WSL to get back running, but! Results on my #845 branch:
image

I think we're in business - there are a few known (sporadic) failures I'll work on next, even on Windows & Appveyor.

**Comment by pruiz** [comment]
@NickCraver Weee!!!! only 5 left! awesome ;)

**Comment by NickCraver** [comment]
**Mono now works on the master branch** using the `net46` DLL as an example (see #845 for the final diff), so I'm going to close this out to cleanup.

Thanks to everyone that pitched in here and especially @pruiz for driving this home. Much appreciated!

There's a lot of work to do on a v2 release but I'll try and get tests green and packages on MyGet as soon as possible. The tests are super tricky, so it'll take a bit.

**Comment by NickCraver** [comment]
I've opened up issues for the items we've hit here: https://github.com/mono/mono/issues/8965 and https://github.com/mono/mono/issues/8967 in hopes it helps someone else down the road.

**Pull request closed without merging**